nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+    h = torch.zeros((mx_int + 1, mx_int + 1))
+    for n in range(num_vox):
+        h[img0[n], mg1[n]] += 1
+
+    Parameters
+    ----------
+    img0 : (X, Y, Z) tensor_like
+        First image volume.
+    img1 : (X, Y, Z) tensor_like
+        Second image volume.
+    fwhm : float, default=7
+        Full-width at half max of Gaussian kernel, for smoothing
+        histogram.
+    mx_int : float, default=255
+        Max intensity of input images.
+
+    Returns
+    ----------
+    H : (mx_int + 1, mx_int + 1) tensor_like
+        Joint intensity histogram.
+
+    """
+    fwhm = (fwhm,) * 2
+
+    # Convert each 'coordinate' of intensities to an index
+    # (replicates the sub2ind function of MATLAB)
+    img0 = img0.flatten().floor()
+    img1 = img1.flatten().floor()
+    sub = torch.stack((img0, img1),
+                      dim=1)  # (num_vox, 2)
+    to_ind = torch.tensor((1, mx_int + 1),
+                          dtype=sub.dtype, device=img0.device)[..., None]  # (2, 1)
+    ind = torch.tensordot(sub, to_ind, dims=([1], [0]))  # (nvox, 1)
+
+    # Build histogram H by adding up counts according to the indicies in ind
+    H = torch.zeros(mx_int + 1, mx_int + 1, device=img0.device, dtype=ind.dtype)
+    H.put_(ind.long(), torch.ones(1, device=img0.device, dtype=ind.dtype).expand_as(ind), accumulate=True)
+
+    # Smooth the histogram
+    smo = smooth(('gauss',) * 2, fwhm=fwhm, device=img0.device, dtype=torch.float32, sep=True)
+    # Padding amount for subsequent convolution
+    size_pad = (smo[0].shape[2], smo[1].shape[3])
+    size_pad = (torch.tensor(size_pad) - 1) // 2
+    size_pad = tuple(size_pad.int().cpu().tolist())
+    # Smooth deformation with Gaussian kernel (by convolving)
+    H = pad(H, size_pad, side='both')
+    H = H[None, None, ...]
+    H = F.conv1d(H, smo[0])
+    H = F.conv1d(H, smo[1])[0, 0, ...]
+
+    # Add eps
+    H = H + 1e-7
+
+    # # Visualise histogram
+    # plt.figure(num=fig_num)
+    # plt.imshow(H,
+    #     cmap='coolwarm', interpolation='nearest',
+    #     aspect='equal', vmax=0.05*H.max())
+    # plt.axis('off')
+    # plt.show()
+
+    return H
+
+
+def _init_torch(opt):
+    """Check device and set some Torch stuff.
+    """
+    opt['device'] = torch.device(opt['device'] if torch.cuda.is_available() else 'cpu')
+    torch.set_grad_enabled(False)
+    torch.backends.cudnn.benchmark = True
+
+    return opt
+
+
+def _optimise(q, Nq, args, opt):
+    """Run optimisation routine to fit q.
+
+    Parameters
+    ----------
+    q : (N, Nq) tensor_like
+        Lie algebra of affine registration fit.
+    Nq : int
+        Number of Lie basis.
+    args : tuple
+        All arguments for optimiser (except the parameters to be fitted)
+    opt : dict
+        See run_affine_reg.
+
+    Returns
+    ----------
+    q : (N, Nq) tensor_like
+        Optimised Lie algebra of affine registration fit.
+
+    """
+    if opt['optimiser'] == 'powell':
```

Comment:
(side thought) I wonder if it's possible to use/derive from the optimizers in pytorch.

Reply:
Could you elaborate maybe?

This is where I was thinking of adding Bayesian optimisation, then it would be very easy to compare performance.

Reply:
It's really focused on gradient-based optimization, but the base class is there: https://pytorch.org/docs/stable/optim.html#torch.optim.Optimizer

Maybe it's not suited to powell. I don't know.
Also, I wonder if we could use autograd to optimise the whole thing with pytorch's optimizers.
Last thing we could check is how things are done in BoTorch (https://botorch.org/docs/design_philosophy)

Reply:
Okay, I have already thought about BoTorch, will test that first I think.